kratos/factories/base_factory.h

```diff
@@ -0,0 +1,236 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Vicente Mataix Ferrandiz
+//
+
+#if !defined(KRATOS_BASE_FACTORY_H_INCLUDED )
+#define  KRATOS_BASE_FACTORY_H_INCLUDED
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/kratos_parameters.h"
+#include "includes/kratos_components.h"
+#include "includes/shared_pointers.h"
+
+namespace Kratos
+{
+///@name Type Definitions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/**
+ * @class FactoryMethods
+ * @ingroup KratosCore
+ * @brief Here we define some common methods
+ * @details Defines the base class factory methods
+ * @author Vicente Mataix Ferrandiz
+ */
+class FactoryMethods
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of FactoryMethods
+    KRATOS_CLASS_POINTER_DEFINITION(FactoryMethods);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /** Constructor.
+     */
+    explicit FactoryMethods(){}
+
+    /** Destructor.
+     */
+    virtual ~FactoryMethods(){}
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief This method checks if the linear class is registered
+     * @param rClassName The nanme of the class
+     * @return True if registered, false otherwise
+     */
+    virtual bool Has(const std::string& rClassName) const
+    {
+        KRATOS_ERROR << "Methods must be implemented in the derived class" << std::endl;
+        return false;
+    }
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    virtual std::string Info() const
+    {
+        return "FactoryMethods";
+    }
+
+    /// Print information about this object.
+    virtual void PrintInfo(std::ostream& rOStream) const
+    {
+        rOStream << this->Info() << std::endl;
+    }
+
+    /// Print object's data.
+    virtual void PrintData(std::ostream& rOStream) const
+    {
+    }
+};
+
+/**
+ * @class Factory
+ * @ingroup KratosCore
+ * @brief Here we add the functions needed for the registration of class
+ * @details Defines the base class factory
+ * @author Vicente Mataix Ferrandiz
+ * @tparam TClass The class to create the factory
+ * @tparam TAuxiliarClass The auxiliar class to create the factory
+ */
+template<typename TClass, typename TAuxiliarClass = TClass>
```

Comment:
I think you can remove `TAuxiliarClass`, it seems to be unused

Reply:
It is used in #3185

Reply:
I checked, it is not used in #3185

Reply:
It was...and now it is a leftover

Reply:
It is used:
![imagen](https://user-images.githubusercontent.com/19991680/92460101-d7d3f680-f1c7-11ea-9b4b-21c3e4c978b1.png)

Reply:
![imagen](https://user-images.githubusercontent.com/19991680/92460140-e4f0e580-f1c7-11ea-961d-f26d2892c231.png)

Reply:
why does the B&S factory (B&S is already templated with LinSolver) have another template on the LinSolverType? This is redundant I would say

Reply:
It is necessary in order to be able to consider the auxiliar classes (if you have a better idea...)

Reply:
The problem is that the variadic template is not almighty and this was the only solution

Reply:
I think there seems to be some confusion on how to use it, see my other comment